aten/src/ATen/native/cuda/RNN.cu

```diff
@@ -0,0 +1,601 @@
+        b2r = F2H(0.0); b2i = F2H(0.0); b2n = F2H(0.0);
+#endif
+      }
+
+      offset = (linearIndex/hsz)*5*hsz+linearIndex%hsz;
+
+      accscalar_t rg, ig, ng;
+
+      rg = sigmoid(H2F(ir) + H2F(hr) + H2F(b1r) + H2F(b2r));
+      ig = sigmoid(H2F(ii) + H2F(hi) + H2F(b1i) + H2F(b2i));
+
+      ng = H2F(in) + H2F(b1n) + rg*( H2F(hn)+H2F(b2n) );
+      ng = THCNumerics<accscalar_t>::tanh(ng);
+      *hy = F2H( ng + ig * ( H2F(hx)-ng ) );
+
+      //SAVE FOR BACKWARDS
+      DEVICE_LINEAR_GET(storage, offset+0*hsz) = F2H(rg);
+      DEVICE_LINEAR_GET(storage, offset+1*hsz) = F2H(ig);
+      DEVICE_LINEAR_GET(storage, offset+2*hsz) = F2H(ng);
+      DEVICE_LINEAR_GET(storage, offset+3*hsz) = hx;
+      DEVICE_LINEAR_GET(storage, offset+4*hsz) = F2H(H2F(hn) + H2F(b2n));
+    }
+}
+
+template <typename scalar_t, typename accscalar_t, typename index_type, int indexing_kind>
+#if __CUDA_ARCH__ >= 350
+__launch_bounds__(32 * 16, 4)
+#endif
+__global__ void gru_cell_backward(
+             TensorInfo<scalar_t, index_type> gradInInput,
+             TensorInfo<scalar_t, index_type> gradInHidden,
+             TensorInfo<scalar_t, index_type> gradOutput,
+             TensorInfo<scalar_t, index_type> gradInputHx,
+             TensorInfo<scalar_t, index_type> storage,
+             index_type hsz,
+             index_type totalElements) {
+  for (index_type linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < totalElements;
+       linearIndex += gridDim.x * blockDim.x) {
+    index_type offset = (linearIndex/hsz)*5*hsz+linearIndex%hsz;
+
+    scalar_t rg = DEVICE_LINEAR_GET(storage, offset+0*hsz);
+    scalar_t ig = DEVICE_LINEAR_GET(storage, offset+1*hsz);
+    scalar_t ng = DEVICE_LINEAR_GET(storage, offset+2*hsz);
+    scalar_t hx = DEVICE_LINEAR_GET(storage, offset+3*hsz);
+    scalar_t hn = DEVICE_LINEAR_GET(storage, offset+4*hsz);
+
+    scalar_t go = DEVICE_LINEAR_GET(gradOutput, linearIndex);
+
+    offset = (linearIndex/hsz)*3*hsz+linearIndex%hsz;
+
+    accscalar_t gig = H2F(go)*( H2F(hx)-H2F(ng) )*( 1-H2F(ig) )*H2F(ig);
+    accscalar_t ghx = H2F(go)*H2F(ig);
+    accscalar_t gin = H2F(go)*( 1-H2F(ig) )*( 1-H2F(ng)*H2F(ng) );
+    accscalar_t ghn = gin * H2F(rg);
+    accscalar_t grg = gin *H2F(hn)*( 1-H2F(rg) )*H2F(rg);
+
+    DEVICE_LINEAR_GET(gradInInput, offset+0*hsz) = F2H(grg);
+    DEVICE_LINEAR_GET(gradInInput, offset+1*hsz) = F2H(gig);
+    DEVICE_LINEAR_GET(gradInInput, offset+2*hsz) = F2H(gin);
+
+    DEVICE_LINEAR_GET(gradInHidden, offset+0*hsz) = F2H(grg);
+    DEVICE_LINEAR_GET(gradInHidden, offset+1*hsz) = F2H(gig);
+    DEVICE_LINEAR_GET(gradInHidden, offset+2*hsz) = F2H(ghn);
+    DEVICE_LINEAR_GET(gradInputHx, linearIndex) = F2H(ghx);
+  }
+}
+
+#undef DEVICE_LINEAR_GET
+#undef DEVICE_BIAS_GET
+#undef H2F
+#undef F2H
+
+} // namespace kernel
+
+template<typename scalar_t, typename index_type>
+void lstm_forward_index(const Tensor& input_gates, const Tensor& hidden_gates,
```

Comment:
Which kernel did this used to be? `LSTMFused_updateOutput`?

Reply:
I think I'll rename it to `lstm_forward_impl`

Reply:
`LSTM_forw_ind_wrap` I guess.